### PR TITLE
Add assembler coding standards

### PR DESCRIPTION
**Pull request by rhness**
Please review these draft coding standards.

### CODE REVIEW

**Review by neilbeesley — APPROVED** [review]

File: Assembler coding standards.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    1. Remove the use of "magic numbers" 
+14. Use meaningful comments 
+15. Use meaningful labels 
+16. Use meaningful DS instructions 
+17. Do not specify BLKSIZE for files
+    1. This allows the system to default to the optimal block size  
+    2. It also allows you to override the block size in the JCL   
+18. Make code reentrant 
+    1. Either get your own storage or use an internal stack  mechanism 
+    2. Split the code and data areas in your code
```
Comment (ghannington) [inline Assembler coding standards.md:33]: 'Split the code and data areas...' should be a separate point.

File: Assembler coding standards.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+21. Align data to boundaries
+    1. At the very least it makes dump reading easier
+    2. Also a performance benefit
+22. Align the objects of EXecute and EXRL instructions to doubleword boundaries.  
+    1. This will prevent us from experiencing minor performance hits when crossing 16-byte boundaries and major performance hits when crossing 256-byte boundaries.  
+23. Align the beginning of performance-critical routines on 32-byte boundaries (see http://publibz.boulder.ibm.com/cgi-bin/bookmgr_OS390/BOOKS/IEA2B170/6.2.3?DT=20070605085148)     
+
+## GenevaERS-specific source code format 
+1. Start code with a TITLE line 
+2. Follow with the standard IBM copyright notice
```
Comment (ghannington) [inline Assembler coding standards.md:49]: This should be something like 'Follow with the Apache License (see CONTRIBUTING.md)'

File: Assembler coding standards.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    2. Also a performance benefit
+22. Align the objects of EXecute and EXRL instructions to doubleword boundaries.  
+    1. This will prevent us from experiencing minor performance hits when crossing 16-byte boundaries and major performance hits when crossing 256-byte boundaries.  
+23. Align the beginning of performance-critical routines on 32-byte boundaries (see http://publibz.boulder.ibm.com/cgi-bin/bookmgr_OS390/BOOKS/IEA2B170/6.2.3?DT=20070605085148)     
+
+## GenevaERS-specific source code format 
+1. Start code with a TITLE line 
+2. Follow with the standard IBM copyright notice 
+3. Follow with the program ID and description 
+4. A Change Control Log is in legacy programs and should not be disturbed, but new programs should not include it.  Developers should refer to RTC for the correct information on program modifications.
```
Comment (ghannington) [inline Assembler coding standards.md:51]: This should be something like 
"A Change Control Log, as is common in legacy programs, is not required. Developers should refer to the history in the git repository for the correct information on program modifications."

File: Assembler coding standards.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+3. Follow with the program ID and description 
+4. A Change Control Log is in legacy programs and should not be disturbed, but new programs should not include it.  Developers should refer to RTC for the correct information on program modifications.  
+
+## GenevaERS-specific guidelines
+1. When a user error occurs, terminate gracefully
+   1. Issue the appropriate return code upon completion 
+      1. 0 - Successful
+      2. 4 - Warning 
+      3. 8 - Error 
+      4. 12 - Severe Error
```
Comment (ghannington) [inline Assembler coding standards.md:59]: Do we want to define the difference between an 'Error' and 'Severe Error'?
We also use RC=16 in cases like failing to create an enclave, or errors in creating and using Pause elements.

File: Assembler coding standards.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+      3. 8 - Error 
+      4. 12 - Severe Error 
+2. Assume that the customer machine supports the instruction set of IBM IBM 3906 (z14, released in July 2017) or later.  
+   1. These instructions are described in z/Architecture Principles of Operation, SA22-7832-11 - Twelfth Edition (August, 2017).
+   2. You can guarantee that your assembly doesn't use any instructions newer than this level by adding the option OPTABLE(ZS8) to your assembly JCL.  
+   3.  Newer instructions can be used, but cannot be the default behavior, and must be paired with a JCL parameter to employ them. 
+3.  Assume that the customer machine is running z/OS 2.4 (released July, 2019) or later.  
+4.  To avoid confusion with the Branch Register instruction, use the J form rather than the BR form of Branch Relative op codes 
+5.  When below-the-bar memory is required, use the STORAGE OBTAIN service rather than the older GETMAIN service. 
+6.  When memory is required, use a *conditional* memory allocation, trap any error, report it to the user, and exit gracefully.
```
Comment (ghannington) [inline Assembler coding standards.md:67]: Should we use the term 'customer' or 'user'?

**Review by ghannington — CHANGES_REQUESTED** [review]
See my comments

File: Assembler coding standards.md (hunk 1 above)
Comment (rhness) [reply Assembler coding standards.md]: Done.

File: Assembler coding standards.md (hunk 2 above)
Comment (rhness) [reply Assembler coding standards.md]: Done.

File: Assembler coding standards.md (hunk 3 above)
Comment (rhness) [reply Assembler coding standards.md]: Done.

**Review by ghannington — APPROVED** [review]

**Pull request merged**